zio-http/src/main/scala/zhttp/service/server/ServerSocketHandler.scala

```diff
@@ -19,13 +20,33 @@ final case class ServerSocketHandler[R, E](
   override def channelRead0(ctx: JChannelHandlerContext, msg: JWebSocketFrame): Unit = {
     WebSocketFrame.fromJFrame(msg) match {
       case Some(frame) =>
-        zExec.unsafeExecute_(ctx) {
-          socket(frame)
+        executeAsync(
+          ctx,
+          ss.onMessage(frame)
             .mapM(frame => ChannelFuture.unit(ctx.writeAndFlush(frame.toJWebSocketFrame)))
-            .runDrain
-        }
+            .runDrain,
+        )
 
       case _ => ()
     }
   }
+
+  def executeAsync(ctx: JChannelHandlerContext, program: ZIO[R, Throwable, Unit]): Unit = {
+    zExec.unsafeExecute(ctx, program) {
+      case Exit.Success(_)     => ()
+      case Exit.Failure(cause) =>
+        cause.failureOption match {
+          case Some(error: Throwable) => ctx.fireExceptionCaught(error)
+          case _                      => ()
+        }
+        ctx.close()
+        ()
+    }
+  }
+
+  override def exceptionCaught(ctx: JChannelHandlerContext, x: Throwable): Unit =
+    executeAsync(ctx, ss.onError(x).uninterruptible)
+
+  override def channelUnregistered(ctx: JChannelHandlerContext): Unit =
+    executeAsync(ctx, ss.onClose(ctx.channel().remoteAddress(), None).uninterruptible)
```

Comment:
Instead of using `onOpen` in RequestHandler can we move it here?

```scala  
  override def channelRegistered(ctx: JChannelHandlerContext): Unit =
    executeAsync(ctx, ss.onOpen(ctx.channel().remoteAddress()).uninterruptible)
```

Reply:
ignore.

Reply:
Actually we I think we can do something like — 

```scala
override def userEventTriggered(ctx: JChannelHandlerContext, event: AnyRef): Unit=
  event match {
    case HandshakeComplete => executeAsync(ctx, ss.onOpen(ctx.channel().remoteAddress()).uninterruptible)
    case event => ctx.fireUserEventTriggered(event)
  }
```